contrib/coredns/policy/policy.go

```diff
@@ -150,7 +196,54 @@ func (p *PolicyMiddleware) ServeDNS(ctx context.Context, w dns.ResponseWriter, r
 	}
 
 	if response.Permit {
-		return middleware.NextOrFailure(p.Name(), p.Next, ctx, w, r)
+
+		lr := new(LocalResponseWriter)
+		lr.localAddr = w.LocalAddr()
+		lr.remoteAddr = w.RemoteAddr()
+		lr.Msg = r
+
+		upstreamResult := new(dns.Msg)
+		upstreamResult = r
+		status,err := middleware.NextOrFailure(p.Name(), p.Next, ctx, lr, upstreamResult)
```

Comment:
`upstreamResult` is not needed. You can just pass `r`. The critical thing is how you have replaced `w` with `lr`.